src/wayland/shell/xdg/mod.rs

```diff
@@ -108,36 +113,270 @@ mod xdg_handlers;
 // compatibility handlers for the zxdg_shell_v6 protocol, its earlier version
 mod zxdgv6_handlers;
 
-/// Metadata associated with the `xdg_surface` role
+macro_rules! xdg_role {
+    ($(#[$role_attr:meta])* $role:ident,
+     $configure:ty,
+     $(#[$attr:meta])* $element:ident {$($(#[$field_attr:meta])* $vis:vis$field:ident:$type:ty),*},
+     $role_ack_configure:expr) => {
+        $(#[$role_attr])*
+        pub type $role = Option<$element>;
```

Comment:
Why is this an Option?

Reply:
The option represents the destruction of the `xdg_role`. This is again checked in the destruction of the `xdg_surface` which raises a protocol error if the role has not be destroyed before destroying the surface.
Before I changed the code this was represented as a `PendingState` enum with `Toplevel`, `Popup` and `None`.
As I have removed the enum I used a `Option` to represent the `None` in both roles.
An alternative could be to have a `is_valid` or `is_destroyed` field in the base attributes and check against that.

Reply:
Right, the protocol specifies the order of destruction. Okay then. :+1: